src/NULS/Address.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright © 2017-2019 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+#include "Address.h"
+
+#include <TrezorCrypto/ecdsa.h>
+
+#include "../Base58.h"
+#include "../BinaryCoding.h"
+#include "../HexCoding.h"
+
+using namespace TW::NULS;
+
+bool Address::isValid(const std::string& string) {
+    if (string.empty()) {
+        return false;
+    }
+    std::string prefix = "NULSd";
+    if (string.length() <=  prefix.length()) {
+        return false;
+    }
+
+    std::string address = string.substr(prefix.length(), string.length() - prefix.length());
+
+    Data decoded = TW::Base58::bitcoin.decode(address);
+    if (decoded.size() != size) {
+        return false;
+    }
+
+    // Check Xor
+    uint8_t checkSum = 0x00;
+    for (int i = 0; i < 23; ++i) {
+        checkSum ^= decoded[i];
+    }
+
+    return decoded[23] == checkSum;
+}
+
+Address::Address(const TW::PublicKey& publicKey) {
+    // Main-Net chainID
+    bytes[0] = 0x01;
+    bytes[1] = 0x00;
+    // Address Type
+    bytes[2] = 0x01;
+
+    ecdsa_get_pubkeyhash(publicKey.bytes.data(), HASHER_SHA2_RIPEMD, bytes.begin() + 3);
+
+    // Calc chechsum
+    uint8_t checkSum = 0x00;
+    for (int i = 0; i < 23; ++i) {
+        checkSum ^= bytes[i];
+    }
+    bytes[23] = checkSum;
+}
+
+Address::Address(const std::string& string) {
+    std::string prefix = "NULSd";
+    std::string address = string.substr(prefix.length(), string.length() - prefix.length());   
+    const auto decoded = Base58::bitcoin.decode(address);
+    if (decoded.size() != Base58Address::size) {
+        throw std::invalid_argument("Invalid address string");
+    }
+
+    std::copy(decoded.begin(), decoded.end(), bytes.begin());
+}
+
+uint16_t Address::chainID() const {
+    return decode16LE(bytes.data());
+}
+
+uint8_t Address::type() const {
+    return bytes[2];
+}
+
+bool Address::isValid() const {
+    if (chainID() != MainNetID) {
+        return false;
+    }
+    if (type() != 0x01 && type() != 0x02) {
+        return false;
+    }
+
+    uint8_t checkSum = 0x00;
+    for (int i = 0; i < 23; ++i) {
+        checkSum ^= bytes[i];
+    }
+
+    return bytes[23] == checkSum;
+}
+
+TW::PrivateKey Address::importHexPrivateKey(std::string hexPrivateKey) {
```

Comment:
remove this, we only support import via recovery phrases

Reply:
reserve it.